### PR TITLE
New Recents API Format

### LINKED ISSUE #41957: Beef up Recents 🥩
related to https://github.com/metabase/metabase/issues/39528
[product doc ](https://www.notion.so/metabase/Add-Recents-tab-to-the-new-EntityPicker-component-895cd80f2b0f4e56a008a82b436c3a00?pm=c)

For the new recent item picker, we're going to need some changes to the activity/recents api

- We need to track the most recent 20 items that each user has "touched" (ie viewed) of each entity type	
	- question
	- model
	- table (probably don't need right now)
	- dashboard
	- collection
- for a collection to be "recently viewed" [means](https://metaboat.slack.com/archives/C064EB1UE5P/p1714504102513489?thread_ts=1714169076.361039&cid=C064EB1UE5P) that:
	- the collection was created
	- a dashboard or question was saved in the collection
	- a dashboard, question, or collection was moved into the collection
- We also need more information about each item than the current api provides
	- `can_write`
	- `last_touched` or some sort of proxy for sorting by most recently used (i think this is what the timestamp field in the recent_views table is doing)
	- parent information - collection name + authority level (eventually parent db for tables)
- While we're at it, can we make the payload for the recents endpoint conform more closely to search and/or collection items?
	- 	nesting `model_object` inside each recent item is a weird pattern that's used nowhere else and is completely unnecessary, we can just populate a `model` property on each item like we do elsewhere
	- we should probably update the popular items api to match

### Desired RecentItem Shape
(open to discussion)

```ts
interface RecentItem {
  id: number;
  name: string;
  model: string; // should differentiate between regular cards and "dataset" for models
  can_write: boolean;
  display: string; // for questions
  parent_collection: { // for non-tables
    id: number;
    name: string;
    authority_level: "official" | null; // for collections
  } | null;
  database: { // for tables
    id: number;
    name: string; // display name
  } | null;
  authority_level: "official" | null; // for collections
  moderated_status: "verified" | null; // for models
  timestamp: string; // last touched
}
```
```[tasklist]
### Tasks
- [x] update return from /activity/recent_views
- [x] track models as distinct from cards
- [x] update recents to preserve inside category (not just last 30 whatevers, but 20 cards, 20 dashboards, etc...)
- [x] Capture collection uses (add event for collection creation, event for moving card, dash, collection)
- [x] update /popular_items return shape to match recent_views
- [x] https://github.com/metabase/metabase/issues/42665
```

### PR DESCRIPTION
**Pull request by escherize**
resolves #41957 

The way this works is by querying `recent_views` for the last 20 of each sort of model. This query joins with report_card so we can differentiate questions (cards) from datasets (models).

Then each one is passed through post-process, a multimethod that dispatches on the model (with a distinction between cards and datasets) which does some classification, and adds details to each RecentViews/Item.

### Tasks
- [x] update return from /activity/recent_views
- [x] track models as distinct from cards
- [x] update recents to preserve inside category (not just last 30 whatevers, but 20 cards, dashboards, models, collections, and tables)
- [x] add collection events for:
    - [x] collection creation, 
    - [x] moving card, 
    - [x] dash, 
    - [x] collection)
- [x] make sure we are pruning `recent_views` properly
- [x] update /popular_items return shape to match recent_views
- [x] tests
- [x] return `moderated_status` on cards, models, (and maybe dashboards)
- [x] filter out archived / inactive values
- [x] add description to all models
- [x] more tests

### CODE REVIEW

**Comment by dpsutton** [comment]
Some comments:

#### A nice ns docstring:

```clojure
(ns metabase.models.recent-views
  "The Recent Views table is used to track the most recent views of objects such as Cards, Models, Tables, Dashboards,
  and Collections for each user. It offers a simple api to add a recent, and fetch the list of recents.

  Fetch recent items: `(recent-view/get-list <user-id>)`
  add recent item: `(recent-views/update-users-recent-views! <user-id> <model> <model-id>)`

  When adding a recent item, duplicates will be removed, and [[*recent-views-stored-per-user-per-model*]] (20
  currently) are kept of each entity type. Ie, if you were to view lots of cards, it would not push collections and
  dashboards out of your recents."
,,,)
```

(note, enumerating the API makes `metabase.models.recent-views/user-recent-views` an awkward function to have around)

#### distinct ids

I see you're doing most of this in SQL. Want to throw out perhaps an easier impl.
```clojure
recent-views=> (->> (set/index (t2/select :model/RecentViews
                                          :user_id 1
                                          {:order-by [[:timestamp :desc]]})
                               [:model :model_id])
                    (mapcat (comp next val))
                    (map :id))
(40
 61
 48
 49
,,,)
```
I think mine is easier to read. And since we always load this all in memory, felt appropriate. Up to you to accept or decline the change.

#### Issue in spec

```diff
@@ -212,7 +228,7 @@
      :model :table
      :can_write (mi/can-write? :model/Table model_id)
      :timestamp (str timestamp)
-     :database {:id (:database-id table)
+     :database {:id (:db_id table)
                 :name (:name table)}}))
 
```

Tables have a `db_id` not a `database-id`. And this is causing
image

File: src/metabase/models/recent_views.clj [hunk 1]
```diff
@@ -94,3 +130,122 @@
...
+     :parent_collection (get-parent-coll (:collection_id dashboard))}))
+
+(defmethod fill-recent-view-info :table [{:keys [_model model_id timestamp]}]
+  (let [table (t2/select-one :model/Table model_id)]
+    {:id model_id
+     :name (:name table)
+     :model :table
+     :can_write (mi/can-write? :model/Table model_id)
+     :timestamp (str timestamp)
+     :database {:id (:database-id table)
```
Comment (escherize) [inline src/metabase/models/recent_views.clj:215]:
```suggestion
     :database {:id (:db_id table)
```

**Comment by iethree** [comment]
I'm getting this error:
```
{
    "cause": "Invalid output: [{:parent_collection {:authority_level [\"should be either official or , got: :official\"]}} {:parent_collection {:authority_level [\"should be either official or , got: :official\"]}} {:parent_collection {:authority_level [\"should be either official or , got: :official\"]}} {:parent_collection {:authority_level [\"should be either official or , got: :official\"]}} {:parent_collection {:authority_level [\"should be either official or , got: :official\"]}} {:database {:id [\"should be an integer, got: nil\"]}} {:parent_collection {:authority_level [\"should be either official or , got: :official\"]}} {:database {:id [\"should be an integer, got: nil\"]}} {:database {:id [\"should be an integer, got: nil\"]}} {:database {:id [\"should be an integer, got: nil\"]}} {:database {:id [\"should be an integer, got: nil\"]}} {:parent_collection {:authority_level [\"should be either official or , got: :official\"]}} nil nil {:parent_collection {:authority_level [\"should be either official or , got: :official\"]}} {:parent_collection {:authority_level [\"should be either official or , got: :official\"]}} {:parent_collection {:authority_level [\"should be either official or , got: :official\"]}} {:parent_collection {:authority_level [\"should be either official or , got: :official\"]}} {:parent_collection {:authority_level [\"should be either official or , got: :official\"]}} {:parent_collection {:authority_level [\"should be either official or , got: :official\"]}} {:parent_collection {:authority_level [\"should be either official or , got: :official\"]}} {:parent_collection {:authority_level [\"should be either official or , got: :official\"]}} {:parent_collection {:authority_level [\"should be either official or , got: :official\"]}} nil nil nil nil {:parent_collection {:authority_level [\"should be either official or , got: :official\"]}} {:parent_collection {:authority_level [\"should be either official or , got: :official\"]}} {:parent_collection {:authority_level [\"should be either official or , got: :official\"]}}]",
    "via": [
        {
            "type": "clojure.lang.ExceptionInfo",
            "message": "Invalid output: [{:parent_collection {:authority_level [\"should be either official or , got: :official\"]}} {:parent_collection {:authority_level [\"should be either official or , got: :official\"]}} {:parent_collection {:authority_level [\"should be either official or , got: :official\"]}} {:parent_collection {:authority_level [\"should be either official or , got: :official\"]}} {:parent_collection {:authority_level [\"should be either official or , got: :official\"]}} {:database {:id [\"should be an integer, got: nil\"]}} {:parent_collection {:authority_level [\"should be either official or , got: :official\"]}} {:database {:id [\"should be an integer, got: nil\"]}} {:database {:id [\"should be an integer, got: nil\"]}} {:database {:id [\"should be an integer, got: nil\"]}} {:database {:id [\"should be an integer, got: nil\"]}} {:parent_collection {:authority_level [\"should be either official or , got: :official\"]}} nil nil {:parent_collection {:authority_level [\"should be either official or , got: :official\"]}} {:parent_collection {:authority_level [\"should be either official or , got: :official\"]}} {:parent_collection {:authority_level [\"should be either official or , got: :official\"]}} {:parent_collection {:authority_level [\"should be either official or , got: :official\"]}} {:parent_collection {:authority_level [\"should be either official or , got: :official\"]}} {:parent_collection {:authority_level [\"should be either official or , got: :official\"]}} {:parent_collection {:authority_level [\"should be either official or , got: :official\"]}} {:parent_collection {:authority_level [\"should be either official or , got: :official\"]}} {:parent_collection {:authority_level [\"should be either official or , got: :official\"]}} nil nil nil nil {:parent_collection {:authority_level [\"should be either official or , got: :official\"]}} {:parent_collection {:authority_level [\"should be either official or , got: :official\"]}} {:parent_collection {:authority_level [\"should be either official or , got: :official\"]}}]",
            "data": {...}
}
```

File: src/metabase/api/activity.clj [hunk 2]
```diff
@@ -165,41 +153,42 @@
...
-  finds anything archived, deleted, etc it can usually still get 5."
-  []
-  ;; we can do a weighted score which incorporates:
-  ;; total count -> higher = higher score
-  ;; recently viewed -> more recent = higher score
-  ;; official/verified -> yes = higher score
+      (map-indexed
+       (fn [recency-pos {:keys [cnt model_object] :as item}]
+         (let [verified-wt 1
+               official-wt 1
```
Comment (escherize) [inline src/metabase/api/activity.clj:159]: I couldn't resist drive-by altering this, but the logic is unrelated and hasn't changed at all.

File: src/metabase/api/activity.clj [hunk 3]
```diff
@@ -165,41 +153,42 @@
...
+                       ;; a long time ago. This may not even make it into the firs 10 items from the query, even though it might be worth showing
+                       (* (/ cnt max-count) views-wt)]]
+           (assoc item :score (double (reduce + scores))))) items))))
+
+(def ^:private model->precedence
+  {"dashboard"  0
+   "card"       1
+   "dataset"    2
+   "table"      3
+   "collection" 4})
```
Comment (escherize) [inline src/metabase/api/activity.clj:181]: I think the map here makes a lot more sense, though. The "indexes" are more explicitly being used now

File: src/metabase/api/activity.clj [hunk 4]
```diff
@@ -209,15 +198,28 @@
...
-         reverse
-         order-items
+         (sort-by
+          ;; sort by model first, and then score when they are the same model
+          (juxt #(-> % :model model->precedence) #(- (% :score))))
          (take 5)
-         (map #(dissoc % :score)))))
+         (map #(-> %
+                   (assoc :timestamp (:max_ts % ""))
+                   recent-views/fill-recent-view-info)))))
```
Comment (noahmoss) [inline src/metabase/api/activity.clj:213]: nit - might be a bit more readable to break this out into two separate maps, instead of nested threading macros? and/or to use a transducer for the whole pipeline

File: src/metabase/api/collection.clj [hunk 5]
```diff
@@ -958,16 +958,17 @@
...
-      Collection
-      (merge
-        {:name        name
-         :description description
-         :authority_level authority_level
-         :namespace   namespace}
-        (when parent_id
-          {:location (collection/children-location (t2/select-one [Collection :location :id] :id parent_id))})))))
+  (u/prog1 (first
+            (t2/insert-returning-instances!
```
Comment (noahmoss) [inline src/metabase/api/collection.clj:962]: can this just use `insert-returning-instance!`?

File: src/metabase/models/recent_views.clj [hunk 6]
```diff
@@ -1,9 +1,22 @@
 (ns metabase.models.recent-views
-  "The Recent Views table is used to track the most recent views of objects such as Cards, Tables and Dashboards for
-  each user."
+  "The Recent Views table is used to track the most recent views of objects such as Cards, Models, Tables, Dashboards,
+  and Collections for each user, for an up to date list, see [[models-of-interest]].
```
Comment (noahmoss) [inline src/metabase/models/recent_views.clj:3]: grammar nit (or could just change the `,` to `;`)
```suggestion
  "The Recent Views table is used to track the most recent views of objects such as Cards, Models, Tables, Dashboards,
  and Collections for each user. For an up to date list, see [[models-of-interest]].
```

File: src/metabase/models/recent_views.clj [hunk 7]
```diff
@@ -79,18 +121,177 @@
...
+  [model]
+  (cond
+    (false? (:archived model)) model
+
+    (true? (:archived model)) nil
+
+    (nil? (:archived model))
+    (throw (ex-info "Archived field is nil" {:model model}))))
+
+(defmethod fill-recent-view-info :card [{:keys [_model model_id timestamp model_object]}]
```
Comment (noahmoss) [inline src/metabase/models/recent_views.clj:203]: Would it make sense for the impls of this multimethod to live in the model namespaces for each model? That way each model is responsible for what information is included in recent views, rather than blowing up this namespace with all the implementations. That's the pattern I followed with `audit-log/model-details` for instance.

Don't need to block on this, just a suggestion but it could go either way.

File: src/metabase/models/recent_views.clj [hunk 8]
```diff
@@ -79,18 +121,177 @@
...
+  (when-let [card (ellide-archived
+                   (or model_object (t2/select-one :model/Card model_id)))]
+    {:id model_id
+     :name (:name card)
+     :description (:description card)
+     :display (when-let [display (:display card)] (name display))
```
Comment (noahmoss) [inline src/metabase/models/recent_views.clj:209]: nit
```suggestion
     :display (some-> (:display card) name)
```

File: test/metabase/models/recent_views_test.clj [hunk 9]
```diff
@@ -4,90 +4,240 @@
    [java-time.api :as t]
    [metabase.models.recent-views :as recent-views]
    [metabase.test :as mt]
+   [metabase.util.log :as log]
    [toucan2.core :as t2]
    [toucan2.tools.with-temp :as t2.with-temp]))
 
+(def test-user :rasta)
```
Comment (noahmoss) [inline test/metabase/models/recent_views_test.clj:11]: is this needed? most tests just use test user names directly inline

**Review by noahmoss — APPROVED** [review]

File: test/metabase/models/recent_views_test.clj (hunk 9 above)
Comment (escherize) [reply test/metabase/models/recent_views_test.clj:11]: Fixed

File: src/metabase/models/recent_views.clj (hunk 8 above)
Comment (escherize) [reply src/metabase/models/recent_views.clj:209]: Good call. I'll go with `(some-> card :display name)` even.

File: src/metabase/models/recent_views.clj (hunk 7 above)
Comment (escherize) [reply src/metabase/models/recent_views.clj:203]: Glad you asked! I thought about that too, but I figured that I wanted these methods to be seen from the POV of a recent view item, instead of a model. Part of the reason for that is the impedence mismatch between :model/Card being both questions and models (and metrics etc if we ever add those?). So I thought it was more simple to handle things as they are on the api side, instead of the way they are on our model/instance side.

File: src/metabase/models/recent_views.clj (hunk 6 above)
Comment (escherize) [reply src/metabase/models/recent_views.clj:3]: Fixed

File: src/metabase/api/collection.clj (hunk 5 above)
Comment (escherize) [reply src/metabase/api/collection.clj:962]: Woah, that's handy. Thanks for pointing that out. Fixed

File: src/metabase/api/activity.clj (hunk 4 above)
Comment (escherize) [reply src/metabase/api/activity.clj:213]: I landed here because I found the transducer style to be harder to read, though it does have the memory pressure benefit. But then in the other direction: for doing 2 maps we'd be creating another intermediate sequence.
